File: app/src/main/java/com/hlibrary/util/ToastUtil.java
package com.hlibrary.util;

import android.content.Context;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.widget.Toast;

import java.lang.ref.WeakReference;

public class ToastUtil {


    private static ToastUtil instance;
    private static WeakReference<Toast> mWeakToast;

    private ToastUtil() {
    }

    public static ToastUtil getInstance() {
        if (instance == null)
            synchronized (ToastUtil.class) {
                if (instance == null)
                    instance = new ToastUtil();
            }
        return instance;
    }

    public synchronized void showToast(@NonNull Context context, @NonNull String text, int duration) {
        Toast mToast = null;
        if (mWeakToast != null)
            mToast = mWeakToast.get();
        if (mToast == null) {
            mToast = Toast.makeText(context, text, duration);
            mWeakToast = null;
        } else {
            mToast.setText(text);
            mToast.setDuration(duration);
        }
        if (mWeakToast == null)
            mWeakToast = new WeakReference<>(mToast);
        mToast.show();
    }

    public synchronized void showToast(@NonNull Context context, @StringRes int resId, int duration) {
        Toast mToast = null;
        if (mWeakToast != null)
            mToast = mWeakToast.get();
        if (mToast == null) {
            mToast = Toast.makeText(context, resId, duration);
            mWeakToast = null;
        } else {
            mToast.setText(resId);
            mToast.setDuration(duration);
        }
        if (mWeakToast == null)
            mWeakToast = new WeakReference<>(mToast);
        mToast.show();
    }

    public static void showLongTime(@NonNull Context context, @NonNull String text) {
        ToastUtil.getInstance().showToast(context, text, Toast.LENGTH_LONG);
    }

    public static void showShortTime(@NonNull Context context, @NonNull String text) {
        ToastUtil.getInstance().showToast(context, text, Toast.LENGTH_SHORT);
    }

    public static void showCustomTime(@NonNull Context context, @NonNull String text, @IntRange(from = 1) int duration) {
        if (duration > 0)
            ToastUtil.getInstance().showToast(context, text, duration);
    }

    public static void showLongTime(@NonNull Context context, @StringRes int resId) {
        ToastUtil.getInstance().showToast(context, resId, Toast.LENGTH_LONG);
    }

    public static void showShortTime(@NonNull Context context, @StringRes int resId) {
        ToastUtil.getInstance().showToast(context, resId, Toast.LENGTH_SHORT);
    }

    public static void showCustomTime(@NonNull Context context, @StringRes int resId, @IntRange(from = 1) int duration) {
        ToastUtil.getInstance().showToast(context, resId, duration);
    }

}

File: app/src/main/java/com/hlibrary/util/ToastUtil.kt
package com.hlibrary.util

import android.content.Context
import android.support.annotation.IntRange
import android.support.annotation.StringRes
import android.widget.Toast

/**
 * @author linwenhui
 */
class ToastUtil private constructor() {

    @Synchronized
    fun showToast(context: Context, text: String, duration: Int) {
        if (mToast == null) {
            mToast = Toast.makeText(context.applicationContext, text, duration)
        } else {
            mToast?.setText(text)
            mToast?.duration = duration
        }
        mToast?.show()
    }

    @Synchronized
    fun showToast(context: Context, @StringRes resId: Int, duration: Int) {
        if (mToast == null) {
            mToast = Toast.makeText(context.applicationContext, resId, duration)

        } else {
            mToast?.setText(resId)
            mToast?.duration = duration
        }
        mToast?.show()
    }

    companion object {


        private var instance: ToastUtil? = null
        private var mToast: Toast? = null

        fun getInstance(): ToastUtil? {
            if (instance == null) {
                synchronized(ToastUtil::class.java) {
                    if (instance == null) {
                        instance = ToastUtil()
                    }
                }
            }
            return instance
        }

        fun showLongTime(context: Context, text: String) {
            ToastUtil.getInstance()?.showToast(context, text, Toast.LENGTH_LONG)
        }

        fun showShortTime(context: Context, text: String) {
            ToastUtil.getInstance()?.showToast(context, text, Toast.LENGTH_SHORT)
        }

        fun showCustomTime(context: Context, text: String, @IntRange(from = 1) duration: Int) {
            if (duration > 0)
                ToastUtil.getInstance()?.showToast(context, text, duration)
        }

        fun showLongTime(context: Context, @StringRes resId: Int) {
            ToastUtil.getInstance()?.showToast(context, resId, Toast.LENGTH_LONG)
        }

        fun showShortTime(context: Context, @StringRes resId: Int) {
            ToastUtil.getInstance()?.showToast(context, resId, Toast.LENGTH_SHORT)
        }

        fun showCustomTime(context: Context, @StringRes resId: Int, @IntRange(from = 1) duration: Int) {
            ToastUtil.getInstance()?.showToast(context, resId, duration)
        }
    }

}
